Bugs/commas in number fields #161196317

fddayan opened the pull request:
This PR is very WIP, these are the items to look at

1. Application new save is not working anymore. It was working with this changes.
1. If you save annual income like $1,000.0 and you refresh the page. You will see 1000.00 in annual income.
1. Improve validation message.

fddayan committed:
SAVEPOINT

fddayan committed:
[#161196317] Validating and parsing annual income

fddayan committed:
Merge branch 'master' into bugs/commas-in-number-fields-#161196317

fddayan committed:
[#161196317] Lint fixed

fddayan committed:
[#161196317] Do not validate anual income if blank

fddayan committed:
[#161196317] Convert to float

callaghanc committed:
[#161196317] Updates regex for annual income to be a whitelist

Also updates related unit test.

callaghanc commented on app/javascript/components/applications/application_form/PaperApplicationForm.js:
```diff
@@ -59,9 +59,22 @@ const buildPrefValidations = (prefs) => {
+    return null
+  } else {
+    if (/[^0-9$,.]/.test(value)) {
+      return 'only accepts numbers, $, commas, and periods.'
```
this error message wording is provisional. i have a question out to design for input on the wording here

callaghanc replied:
resolved with new wording provided by josh

callaghanc commented:
Since @fddayan is ramping down from DAHLIA, I'll take over from him on this PR. He is still available for questions until the 26th though!

Remaining items:
- [x] Look into application save issue mentioned by Fed above (I believe he may having been seeing the expected error message when viewing a newly added application on a Lease Up listing - that throws an error because the new app doesn't have a snapshot. No error occurs when adding and viewing a new app on a non-Lease Up listing).
- [x] Get input from design on Annual Income field error message wording.
- [x] Get decision from design and product on whether to add complete currency parsing in this story.
- [x] Add unit test for the error state of the Annual Income field in the paper application form.

callaghanc committed:
[#161196317] Updates paper app annual income field error msg

callaghanc committed:
[#161196317] Adds unit test for validation of annual income field on paper app

callaghanc commented on spec/javascript/components/applications/application_form/__snapshots__/PaperApplicationForm.test.js.snap:
```diff
@@ -0,0 +1,7290 @@
+                        "lottery_number_manual": null,
+                        "monthly_income": null,
+                        "name": "APP-00191270",
+                        "preferences": Arra
```
this error class on the form group is the beginning of the error state we expect to see when the annual income field is in an invalid state. please examine the following elements to note several more instances where error classes are applied. also notice the presence of the small error message below the field. there will be another comment below this marking the end of the section that is relevant to the error state of the annual income field.

callaghanc replied:
this ends the section of the snapshot that deals with the error state of the annual income field

akegan commented:
@callaghanc looks like this PR has a couple of small merge conflicts. after they're resolved let me know and I can review/approve!

akegan committed:
Merge branch 'master' into bugs/commas-in-number-fields-#161196317

akegan commented:
@callaghanc I went ahead and resolved the merge conflicts since you have such limited hours this week.

akegan commented on spec/javascript/components/applications/application_form/PaperApplicationForm.test.js:
```diff
@@ -0,0 +1,41 @@
+
+      await wait(100)
+
+      expect(wrapper).toMatchSnapshot()
```
@callaghanc and @chrisdolendo Unrelated to this PR, but a question I have in general: Would it be possible to do the same sort of checks that we do in e2e tests for the presence of e.g. a validation error in these snapshot tests?

For example, in the ApplicationNewPage "should fail if required fields are missing" test, we have this expectation. 
```
const errors = await page.$$eval('.form-group.error span.error', divs => divs.map(d => d.textContent))
expect(errors).toContain('Please enter a First Name')
```

we could add something like that in addition to checking that it matches the snapshot, to be a little more explicit about what we're expecting.

callaghanc replied:
already chatted about this in slack, but check out the new test updates i added that do something very similar to this

akegan committed:
[#161196317] Update snapshots after merge from master

callaghanc committed:
[#161196317] Updates unit test for annual income field validation

We had been using a snapshot of the entire PaperApplicationForm to check that the annual income field received the correct error validation. This commit does away with that, saving us a few thousand lines of snapshot, and instead checks the PaperApplicationForm component wrapper directly for the presence of the annual income field validation states.

callaghanc committed:
[#161196317] Updates attrs on modals in snapshots

This reverses a previous update to the snapshotted modals because that update was found not to work in all cases.

callaghanc commented:
@akegan Tests passed on Semaphore! I believe this should be ready for a final PR review from you.

akegan reviewed (approved):
New tests strategy is great. I'll try to incorporate it into my future PRs
